File: motifs/python/substrate_cell.py
"""The substrate cell motif in Python (canonical)."""
from dataclasses import dataclass, field
from typing import Any
import hashlib


@dataclass
class SubstrateCell:
    """The irreducible unit. ID, prev_hash, content_hash, state, witnesses."""
    id: str
    prev_hash: str  # hex
    state: dict
    source: str
    conversation_id: str = 'default'
    witnesses: list = field(default_factory=list)
    
    @property
    def hash(self) -> str:
        """Content hash (FNV-1a 64-bit, matches fleet canary)."""
        canonical = f"{self.id}|{self.conversation_id}|{self.prev_hash}|{self.source}|{json.dumps(self.state, sort_keys=True)}"
        return self._fnv1a_64(canonical)
    
    @staticmethod
    def _fnv1a_64(s: str) -> str:
        h = 0xcbf29ce484222325
        for byte in s.encode('utf-8'):
            h ^= byte
            h = (h * 0x100000001b3) & 0xffffffffffffffff
        return f'0x{h:016x}'


import json

File: motifs/javascript/substrate_cell.js
// The substrate cell motif in JavaScript (TypeScript-style).
export class SubstrateCell {
    constructor(id, prevHash, source, state, conversationId = 'default') {
        this.id = id;
        this.prevHash = prevHash;
        this.source = source;
        this.state = state;
        this.conversationId = conversationId;
        this.witnesses = [];
    }
    
    hash() {
        // FNV-1a 64-bit
        const canonical = `${this.id}|${this.conversationId}|${this.prevHash}|${this.source}|${JSON.stringify(this.state)}`;
        let h = 0xcbf29ce484222325n;
        const prime = 0x100000001b3n;
        for (const byte of new TextEncoder().encode(canonical)) {
            h ^= BigInt(byte);
            h = (h * prime) & 0xffffffffffffffffn;
        }
        return `0x${h.toString(16).padStart(16, '0')}`;
    }
}

export function fnv1a_64(s) {
    let h = 0xcbf29ce484222325n;
    const prime = 0x100000001b3n;
    for (const byte of new TextEncoder().encode(s)) {
        h ^= BigInt(byte);
        h = (h * prime) & 0xffffffffffffffffn;
    }
    return `0x${h.toString(16).padStart(16, '0')}`;
}

// Canary test
const TEST_CANARY = fnv1a_64('café Δ 日本語');
console.assert(TEST_CANARY === '0x024a555471370b18d', 'FNV-1a canary mismatch');
